File: apps/macos/Sources/OpenSeerGUI/AgentdClient.swift
import Darwin
import Foundation

/// Connection to the `openseer agentd` WebSocket daemon. Phase 1
/// (skeleton): reads the rendezvous file the daemon writes on
/// startup, opens a WebSocket, completes the token auth handshake,
/// exposes a request/response API and an event stream.
///
/// Not yet wired into RunSession / voice orb / main composer — those
/// migrations are Phase 2+. For now `runProbe()` validates the full
/// roundtrip end-to-end.
@MainActor
final class AgentdClient: NSObject {
    static let shared = AgentdClient()

    // MARK: - Rendezvous (written by the Python daemon)

    private struct Rendezvous: Codable {
        let host: String
        let port: Int
        let token: String
        let pid: Int
        let started_at: Double
        let protocol_version: Int
    }

    static func rendezvousPath() -> URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".openseer/agentd.json")
    }

    // MARK: - State

    private var task: URLSessionWebSocketTask?
    private var session: URLSession?
    private var rendezvous: Rendezvous?

    /// Server→client replies keyed by request_id. The send side
    /// installs a continuation; the receive loop pops + resumes.
    private var pendingRequests: [String: CheckedContinuation<[String: Any], Error>] = [:]

    /// Server→client event stream keyed by run_id. start_task
    /// callers register a handler; the receive loop calls it for
    /// every event tagged with that run_id.
    private var eventHandlers: [String: ([String: Any]) -> Void] = [:]

    /// Events that arrived for a run_id before its handler was
    /// registered. This is essential: the daemon can (and for the
    /// Phase 1 stub does) emit `task_started` before the caller
    /// gets the start_task ack and calls observe() — without
    /// buffering, the dispatcher would log those as unrouted and
    /// drop them. Replayed in order on observe().
    private var pendingEvents: [String: [[String: Any]]] = [:]

    /// Run IDs we've seen a `task_finished` / `task_failed` for
    /// before any handler was registered. When observe() is later
    /// called for one of these IDs, we replay the buffer and then
    /// immediately tear down the handler since no more events are
    /// coming.
    private var pendingTerminated: Set<String> = []

    /// run_id → handler that resolves an ask_user round-trip. The
    /// handler receives the parsed server payload (question, kind,
    /// options, request_id) and returns the user's reply
    /// asynchronously. AgentdClient sends `user_reply` to the
    /// daemon with that string.
    var askUserHandler: ((AskUserRequest) async -> String?)? = nil

    struct AskUserRequest: Equatable {
        let runId: String
        let requestId: String
        let question: String
        let kind: String         // "confirm" | "choose" | "text"
        let options: [String]
        let attachments: [String]
    }

    private var nextRequestId = 0

    /// Lazily-spawned `openseer agentd` subprocess, kept so we can
    /// later add a "Stop daemon" gesture if we want. Nil after the
    /// rendezvous file is owned by some other (e.g. user-launched)
    /// agentd we attached to.
    private var spawnedDaemon: Process?

    enum AgentdError: Error, CustomStringConvertible {
        case noRendezvous(String)
        case connectionClosed(String)
        case authFailed(String)
        case badResponse(String)

        var description: String {
            switch self {
            case .noRendezvous(let m): return "no rendezvous: \(m)"
            case .connectionClosed(let m): return "connection closed: \(m)"
            case .authFailed(let m): return "auth failed: \(m)"
            case .badResponse(let m): return "bad response: \(m)"
            }
        }
    }

    // MARK: - Connect / auth

    private func loadRendezvous() throws -> Rendezvous {
        let url = Self.rendezvousPath()
        guard FileManager.default.fileExists(atPath: url.path) else {
            throw AgentdError.noRendezvous(
                "\(url.path) missing — start `openseer agentd` first")
        }
        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode(Rendezvous.self, from: data)
    }

    /// Open the WebSocket and complete the auth handshake. Safe to
    /// call when already connected (no-op).
    func connect() async throws {
        if task != nil { return }
        let r = try loadRendezvous()
        rendezvous = r
        guard let url = URL(string: "ws://\(r.host):\(r.port)") else {
            throw AgentdError.connectionClosed("bad url")
        }
        let cfg = URLSessionConfiguration.default
        let session = URLSession(configuration: cfg)
        self.session = session
        let task = session.webSocketTask(with: url)
        self.task = task
        task.resume()

        // Receive loop is detached from connect()'s lifetime so we
        // keep draining messages even after this function returns.
        Task { @MainActor [weak self] in
            await self?.receiveLoop()
        }

        let ack = try await sendRequest(
            "auth", payload: ["token": r.token])
        guard ack["type"] as? String == "ack" else {
            throw AgentdError.authFailed("\(ack)")
        }
        NSLog("[agentd] connected on :%d (protocol v%d)",
              r.port, r.protocol_version)
    }

    func disconnect() {
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
        session?.invalidateAndCancel()
        session = nil
        // Fail any outstanding requests so callers don't hang.
        for (_, cont) in pendingRequests {
            cont.resume(throwing: AgentdError.connectionClosed("disconnect()"))
        }
        pendingRequests.removeAll()
        eventHandlers.removeAll()
        pendingEvents.removeAll()
        pendingTerminated.removeAll()
    }

    // MARK: - Send / receive

    /// Send `{type, request_id, ...payload}`, await the matching
    /// ack/error/pong reply. Returns the full reply dict.
    @discardableResult
    func sendRequest(_ type: String,
                     payload: [String: Any] = [:]) async throws -> [String: Any] {
        guard let task = self.task else {
            throw AgentdError.connectionClosed("not connected")
        }
        nextRequestId += 1
        let rid = "r\(nextRequestId)"
        var msg = payload
        msg["type"] = type
        msg["request_id"] = rid
        let data = try JSONSerialization.data(
            withJSONObject: msg, options: [.withoutEscapingSlashes])
        guard let text = String(data: data, encoding: .utf8) else {
            throw AgentdError.badResponse("non-utf8 payload")
        }
        return try await withCheckedThrowingContinuation { cont in
            pendingRequests[rid] = cont
            task.send(.string(text)) { [weak self] error in
                guard let error else { return }
                Task { @MainActor [weak self] in
                    if let c = self?.pendingRequests.removeValue(forKey: rid) {
                        c.resume(throwing: error)
                    }
                }
            }
        }
    }

    /// Register a handler for events tagged with `run_id`. If
    /// events arrived for this run_id before this call (very common
    /// — the daemon often emits `task_started` between the
    /// start_task ack and the caller's observe()), they're replayed
    /// here in order.
    ///
    /// The handler is NOT removed when `task_finished` / `task_failed`
    /// fires. Post-run events still belong to this run by run_id
    /// (the reflection pass runs in `on_run_end` and emits
    /// `skill_proposed` shortly after `task_finished`, and the
    /// matching `skill_applied` / `skill_discarded` come back from
    /// the daemon on user action). Auto-removing here meant those
    /// trailing events landed in `pendingEvents` and never reached
    /// the UI; the lesson chip silently never appeared.
    ///
    /// Memory: handlers persist for the life of this AgentdClient.
    /// A long-running session that fires hundreds of runs accumulates
    /// closures, but each is small and the entries get evicted on
    /// disconnect. Future: add explicit `unobserve(runId:)` when
    /// RunSession objects are torn down.
    func observe(runId: String,
                 handler: @escaping ([String: Any]) -> Void) {
        eventHandlers[runId] = handler
        if let buffered = pendingEvents.removeValue(forKey: runId) {
            for msg in buffered {
                handler(msg)
            }
        }
        // We still want to clear the terminated-flag bookkeeping
        // entry; it's only used to detect "terminator landed before
        // observe registered" and a re-observe after that is
        // pointless anyway — but the handler itself stays.
        pendingTerminated.remove(runId)
    }

    private func receiveLoop() async {
        while let task = self.task {
            do {
                let msg = try await task.receive()
                let text: String
                switch msg {
                case .string(let s): text = s
                case .data(let d): text = String(data: d, encoding: .utf8) ?? ""
                @unknown default: text = ""
                }
                guard let data = text.data(using: .utf8),
                      let obj = try? JSONSerialization.jsonObject(with: data)
                        as? [String: Any] else {
                    NSLog("[agentd] non-json msg: %@", text)
                    continue
                }
                dispatch(obj)
            } catch {
                NSLog("[agentd] receive failed: %@", "\(error)")
                // Resume outstanding continuations with the failure
                // so callers don't hang waiting on a dead socket.
                for (_, cont) in pendingRequests {
                    cont.resume(throwing: AgentdError.connectionClosed("\(error)"))
                }
                pendingRequests.removeAll()
                self.task = nil
                return
            }
        }
    }

    private func dispatch(_ msg: [String: Any]) {
        let type = msg["type"] as? String ?? ""
        if type == "ask_user" {
            handleAskUser(msg)
            return
        }
        if type == "event", let runId = msg["run_id"] as? String {
            let inner = msg["event"] as? [String: Any]
            let et = inner?["type"] as? String
            let terminator = (et == "task_finished" || et == "task_failed")
            if let handler = eventHandlers[runId] {
                handler(msg)
                // Intentionally NOT removing on terminator — the
                // run's reflection pass fires `skill_proposed` from
                // `on_run_end`, which the agent emits AFTER
                // `task_finished`. Tearing the handler down here
                // routed those trailing events into `pendingEvents`
                // forever (nobody calls observe() again), so the
                // lesson chip never appeared. See observe() doc for
                // the new lifecycle.
            } else {
                // No handler registered yet — buffer for replay
                // when observe(runId:) eventually catches up.
                pendingEvents[runId, default: []].append(msg)
                if terminator {
                    pendingTerminated.insert(runId)
                }
            }
            return
        }
        if let rid = msg["request_id"] as? String,
           let cont = pendingRequests.removeValue(forKey: rid) {
            cont.resume(returning: msg)
            return
        }
        NSLog("[agentd] unrouted msg: %@", "\(msg)")
    }

    // MARK: - Probe (Task #1 deliverable)

    // MARK: - High-level API

    /// Ensure agentd is running and we have a live, authed
    /// connection. Will (a) reuse an existing daemon if the
    /// rendezvous responds, or (b) spawn `openseer agentd` if not.
    /// Pass the resolved `openseer` binary path (same one
    /// `OpenSeerEnv.binaryPath` exposes).
    func ensureRunning(binary: String,
                       startupTimeout: TimeInterval = 10) async throws {
        // 1. If we're already connected and the socket is alive,
        // we're done.
        if task != nil {
            do {
                _ = try await sendRequest("ping", payload: [:])
                return
            } catch {
                // Stale connection; reset and retry.
                disconnect()
            }
        }
        // 2. Rendezvous file present → try connecting to whatever
        // daemon wrote it. If it doesn't answer the ping, treat the
        // file as stale and overwrite by spawning fresh.
        if FileManager.default.fileExists(
            atPath: Self.rendezvousPath().path) {
            do {
                try await connect()
                _ = try await sendRequest("ping", payload: [:])
                return
            } catch {
                NSLog("[agentd] stale rendezvous, respawning: %@",
                      "\(error)")
                disconnect()
                try? FileManager.default.removeItem(
                    at: Self.rendezvousPath())
            }
        }
        // 3. Spawn a fresh agentd. We don't supervise it past this
        // process's lifetime — if the user quits OpenSeerGUI the
        // daemon also goes. Future: detach via launchctl when we
        // want it to keep serving telegram in the background.
        try spawnDaemon(binary: binary)

        // 4. Wait for the rendezvous file to appear, then connect.
        let deadline = Date().addingTimeInterval(startupTimeout)
        while Date() < deadline {
            if FileManager.default.fileExists(
                atPath: Self.rendezvousPath().path) {
                try await connect()
                _ = try await sendRequest("ping", payload: [:])
                return
            }
            try await Task.sleep(nanoseconds: 120_000_000)
        }
        throw AgentdError.connectionClosed(
            "daemon failed to start within \(Int(startupTimeout))s")
    }

    private func spawnDaemon(binary: String) throws {
        // Singleton invariant: only one `openseer agentd` may run at
        // a time. Multiple daemons each take their own AX /
        // screen-recording locks and serialize on macOS's default 30s
        // messaging timeout, surfacing as exactly-30s prep delay per
        // agent step. Best-effort SIGTERM every other matching
        // process before we spawn fresh; we don't wait for the OS
        // to confirm exits because the rendezvous-file poll in
        // `ensureRunning` already handles the startup race.
        terminateStaleDaemons()

        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: binary)
        proc.arguments = ["agentd"]
        var env = ProcessInfo.processInfo.environment
        env["PYTHONUNBUFFERED"] = "1"
        proc.environment = env

        // Pipe daemon stderr/stdout to a rotating log so users can
        // tail it if something explodes; otherwise daemon noise
        // would vanish.
        let logURL = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".openseer/agentd.log")
        FileManager.default.createFile(
            atPath: logURL.path, contents: Data())
        let logHandle = try FileHandle(forWritingTo: logURL)
        proc.standardOutput = logHandle
        proc.standardError = logHandle

        try proc.run()
        spawnedDaemon = proc
        NSLog("[agentd] spawned pid=%d, log=%@",
              proc.processIdentifier, logURL.path)
    }

    /// Find every `openseer.cli agentd` python process and SIGTERM it.
    /// Called before spawning to keep at most one daemon alive —
    /// duplicates cause the AX-timeout contention described above.
    /// pgrep is shipped with macOS and matches against the full
    /// command line via `-f`, which catches our python -m form.
    private func terminateStaleDaemons() {
        let pgrep = Process()
        pgrep.executableURL = URL(fileURLWithPath: "/usr/bin/pgrep")
        // Match both forms the daemon can be launched as:
        //   • console-script  → `/.../bin/openseer agentd`
        //   • module entry    → `python3 -m openseer.cli agentd`
        // Anchoring on `agentd$` keeps us from matching the macOS
        // helpers `talagentd` / `gamecontrolleragentd` (no `openseer`
        // in their argv anyway) and from matching incidental uses of
        // the word "agentd" mid-command. Case-sensitive by design so
        // the GUI binary itself (`OpenSeerGUI`) doesn't match.
        pgrep.arguments = ["-f", "openseer.* agentd$"]
        let pipe = Pipe()
        pgrep.standardOutput = pipe
        pgrep.standardError = Pipe()    // swallow noise
        do {
            try pgrep.run()
            pgrep.waitUntilExit()
        } catch {
            NSLog("[agentd] singleton: pgrep failed: %@", "\(error)")
            return
        }
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        let text = String(data: data, encoding: .utf8) ?? ""
        let selfPid = ProcessInfo.processInfo.processIdentifier
        let pids: [pid_t] = text.split(separator: "\n").compactMap {
            pid_t($0.trimmingCharacters(in: .whitespaces))
        }.filter { $0 != selfPid && $0 > 0 }
        for pid in pids {
            let rc = kill(pid, SIGTERM)
            NSLog("[agentd] singleton: SIGTERM stale daemon pid=%d (rc=%d)",
                  pid, rc)
        }
        // Tiny pause so the doomed daemons get a chance to release
        // their AX / screen-recording locks before the fresh one
        // grabs them. Empirically 150ms is plenty; we don't block
        // longer because the rendezvous poll covers startup races
        // and waitpid-style sync would need us to be each daemon's
        // parent (most aren't, e.g. CLI-started ones reparent to
        // launchd).
        if !pids.isEmpty {
            usleep(150_000)
        }
    }

    /// Persist the SKILL.md the daemon's reflection pass parked at
    /// `~/.openseer/runs/<runId>/proposed_skill.md`. The body itself
    /// is on disk on the daemon side; this request just authorizes
    /// the write. Returns the path the daemon wrote to on success.
    @discardableResult
    func applySkill(runId: String) async throws -> String {
        let ack = try await sendRequest(
            "apply_skill", payload: ["run_id": runId])
        guard let path = ack["skill_path"] as? String else {
            throw AgentdError.badResponse(
                "apply_skill ack missing skill_path: \(ack)")
        }
        return path
    }

    /// Tell the daemon the user rejected the proposed skill — it
    /// unlinks the sidecar so the suggestion can't be re-applied
    /// after dismissal and emits a `skill_discarded` event so any
    /// stale UI chips clear themselves.
    func discardSkill(runId: String) async throws {
        _ = try await sendRequest(
            "discard_skill", payload: ["run_id": runId])
    }

    /// Start a task on the daemon. Returns the run_id immediately
    /// after the ack; events flow through `onEvent` as the agent
    /// produces them. `onEvent` will see at least one terminator
    /// (`task_finished` or `task_failed`) at the end.
    @discardableResult
    func startTask(prompt: String,
                   dryRun: Bool,
                   sessionContext: String?,
                   backgroundMode: Bool = false,
                   onEvent: @escaping ([String: Any]) -> Void)
        async throws -> String {
        var payload: [String: Any] = [
            "task": prompt,
            "dry_run": dryRun,
            "background_mode": backgroundMode,
        ]
        if let ctx = sessionContext, !ctx.isEmpty {
            payload["session_context"] = ctx
        }
        let ack = try await sendRequest("start_task", payload: payload)
        guard let runId = ack["run_id"] as? String else {
            throw AgentdError.badResponse(
                "start_task ack missing run_id: \(ack)")
        }
        // Register the observer *after* the ack: any events the
        // daemon emitted between our send and now are buffered in
        // `pendingEvents` and replayed by observe().
        observe(runId: runId, handler: onEvent)
        return runId
    }

    /// Cancel a running task. Daemon writes the CANCEL sentinel +
    /// asyncio-cancels the runner; the agent loop emits its own
    /// task_finished(status="interrupted") in response.
    func cancelTask(runId: String) async {
        do {
            _ = try await sendRequest(
                "cancel_task", payload: ["run_id": runId])
        } catch {
            NSLog("[agentd] cancel failed: %@", "\(error)")
        }
    }

    /// Hand-off: suspend the agent so the user can drive the
    /// mouse/keyboard for a few steps. Agent re-reads AX/screen
    /// state when resumed, so it picks up whatever changed.
    func holdTask(runId: String) async {
        do {
            _ = try await sendRequest(
                "hold_task", payload: ["run_id": runId])
        } catch {
            NSLog("[agentd] hold failed: %@", "\(error)")
        }
    }

    func resumeTask(runId: String) async {
        do {
            _ = try await sendRequest(
                "resume_task", payload: ["run_id": runId])
        } catch {
            NSLog("[agentd] resume failed: %@", "\(error)")
        }
    }

    /// Handle an incoming ask_user from the daemon. Routes to the
    /// askUserHandler, awaits its return, and posts a user_reply
    /// back. If no handler is registered we reply with null which
    /// the agent treats as "user didn't respond" and terminates.
    private func handleAskUser(_ msg: [String: Any]) {
        guard let runId = msg["run_id"] as? String,
              let reqId = msg["request_id"] as? String else {
            NSLog("[agentd] malformed ask_user: %@", "\(msg)")
            return
        }
        let req = AskUserRequest(
            runId: runId,
            requestId: reqId,
            question: msg["question"] as? String ?? "",
            kind: msg["kind"] as? String ?? "text",
            options: msg["options"] as? [String] ?? [],
            attachments: msg["attachments"] as? [String] ?? []
        )
        guard let handler = askUserHandler else {
            NSLog("[agentd] no askUserHandler — replying null for %@",
                  req.requestId)
            sendUserReply(requestId: req.requestId, reply: nil)
            return
        }
        Task { @MainActor [weak self] in
            let reply = await handler(req)
            self?.sendUserReply(requestId: req.requestId, reply: reply)
        }
    }

    private func sendUserReply(requestId: String, reply: String?) {
        // `ask_id` carries the original ask_user's request_id;
        // `request_id` on this user_reply message is allocated by
        // sendRequest for the ack flow. Two distinct ids to avoid
        // colliding the ask_user round-trip with the user_reply's
        // own ack.
        Task { @MainActor [weak self] in
            do {
                var payload: [String: Any] = ["ask_id": requestId]
                if let reply { payload["reply"] = reply }
                else        { payload["reply"] = NSNull() }
                _ = try await self?.sendRequest(
                    "user_reply", payload: payload)
            } catch {
                NSLog("[agentd] sending user_reply failed: %@",
                      "\(error)")
            }
        }
    }

    // MARK: - Smoke test

    /// Smoke test: connect, ping, start the stubbed task, wait for
    /// its two events, disconnect. Logs each step via NSLog so the
    /// run shows up in Console.app under process == OpenSeerGUI.
    func runProbe() async {
        NSLog("[agentd] probe: starting")
        do {
            try await connect()
            let pong = try await sendRequest(
                "ping", payload: ["payload": "hi"])
            NSLog("[agentd] probe: pong=%@", "\(pong)")

            let ack = try await sendRequest(
                "start_task", payload: ["task": "probe"])
            guard let runId = ack["run_id"] as? String else {
                NSLog("[agentd] probe: ack missing run_id: %@", "\(ack)")
                return
            }
            NSLog("[agentd] probe: ack run_id=%@", runId)

            // Collect both stub events. The receive loop pushes them
            // into our handler; we resume the continuation when the
            // terminator (`task_finished`) arrives.
            await withCheckedContinuation { (cont: CheckedContinuation<Void, Never>) in
                var finished = false
                observe(runId: runId) { msg in
                    guard !finished else { return }
                    let event = msg["event"] as? [String: Any]
                    let et = event?["type"] as? String ?? "?"
                    NSLog("[agentd] probe: event %@", et)
                    if et == "task_finished" || et == "task_failed" {
                        finished = true
                        cont.resume()
                    }
                }
            }
            NSLog("[agentd] probe: PASS")
        } catch {
            NSLog("[agentd] probe: FAIL — %@", "\(error)")
        }
    }
}

File: openseer/agentd_client.py
"""Sync WebSocket client for `openseer agentd`.

Used by:
- `openseer task` CLI — attach to a running daemon when one exists,
  so manual `openseer task "..."` invocations share state with the
  GUI session and skip the cold Python import every run. Falls back
  cleanly to direct `agent.run()` if no daemon is reachable.

- (future) the Telegram daemon — route inbound chat messages through
  the same agentd instance the GUI uses, so all clients see the
  same ask_user / hand-off / event stream surface.

If the daemon isn't running, `try_open()` returns None and the
caller is expected to fall back to direct in-process execution.
"""
from __future__ import annotations

import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterator, Optional

import websockets.exceptions
from websockets.sync.client import connect


_RENDEZVOUS_PATH = Path.home() / ".openseer" / "agentd.json"


@dataclass
class _Pending:
    """One in-flight server-initiated round-trip (currently only
    ask_user) — we send a reply once stdin produces an answer."""
    request_id: str
    payload: dict[str, Any]


class AgentdClient:
    """Sync agentd client. Construct with `try_open()` (which connects +
    auths) and call `close()` when done. Don't use `with` —
    `try_open` already entered the connection."""

    def __init__(self, host: str, port: int, token: str):
        self.url = f"ws://{host}:{port}"
        self.token = token
        self.ws: Any = None
        self._rid = 0

    def connect_and_auth(self) -> None:
        """Called once by try_open(). Raises on failure; caller
        catches and falls back to the direct path."""
        self.ws = connect(self.url, open_timeout=3.0)
        rid = self._send({"type": "auth", "token": self.token})
        ack = self._recv()
        if ack.get("type") != "ack" or ack.get("request_id") != rid:
            raise ConnectionError(f"agentd auth failed: {ack}")

    def close(self) -> None:
        if self.ws is not None:
            try:
                self.ws.close()
            except Exception:
                pass
            self.ws = None

    # ─── low-level ──────────────────────────────────────────────────

    def _next_rid(self) -> str:
        self._rid += 1
        return f"r{self._rid}"

    def _send(self, msg: dict[str, Any]) -> str:
        rid = self._next_rid()
        out = {**msg, "request_id": rid}
        self.ws.send(json.dumps(out, ensure_ascii=False))
        return rid

    def _recv(self) -> dict[str, Any]:
        raw = self.ws.recv()
        return json.loads(raw)

    # ─── high-level ─────────────────────────────────────────────────

    def run_task(self, task: str, *, dry_run: bool,
                 max_steps: int = 200,
                 session_context: str = "",
                 ask_user_via_stdin: bool = True
                 ) -> Iterator[dict[str, Any]]:
        """Send start_task, yield each event dict as it arrives.

        Yields a synthetic `{"type": "_ack", "run_id": ...}` first.
        Continues until the agent emits task_finished or task_failed,
        then returns.

        If the agent calls ask_user, this generator interactively
        prompts on stdin and posts user_reply automatically. Set
        `ask_user_via_stdin=False` to instead reply None (timeout
        path) — useful in scripts where there's no human.
        """
        msg: dict[str, Any] = {
            "type": "start_task",
            "task": task,
            "dry_run": dry_run,
            "max_steps": max_steps,
        }
        if session_context:
            msg["session_context"] = session_context
        sent_rid = self._send(msg)

        run_id: Optional[str] = None

        while True:
            reply = self._recv()
            t = reply.get("type")

            # Start-task ack (must match our request_id).
            if (t == "ack" and reply.get("request_id") == sent_rid
                    and "run_id" in reply):
                run_id = reply["run_id"]
                yield {"type": "_ack", "run_id": run_id}
                continue

            # Server-initiated ask_user — prompt stdin synchronously
            # and post user_reply back.
            if t == "ask_user":
                self._handle_ask_user(reply, via_stdin=ask_user_via_stdin)
                continue

            # Events for our run.
            if t == "event" and reply.get("run_id") == run_id:
                ev = reply.get("event") or {}
                yield ev
                et = ev.get("type")
                if et in ("task_finished", "task_failed"):
                    return

            # Anything else (ack for our user_reply, errors) — ignore.

    def _handle_ask_user(self, msg: dict[str, Any], *,
                         via_stdin: bool) -> None:
        ask_id = msg.get("request_id") or ""
        question = msg.get("question") or ""
        kind = msg.get("kind") or "text"
        options = msg.get("options") or []

        reply: Optional[str]
        if not via_stdin:
            reply = None
        else:
            reply = _prompt_user_for(question, kind, options)

        out: dict[str, Any] = {"type": "user_reply", "ask_id": ask_id}
        out["reply"] = reply if reply is not None else None
        # Don't await ack (we'd block the event stream). Daemon's
        # ack flows back into the main loop and is ignored there.
        self._send(out)


def _prompt_user_for(question: str, kind: str,
                     options: list[str]) -> Optional[str]:
    """Block on stdin for a reply. Returns the chosen / typed
    string, or None if the user pressed Ctrl+D or just enter."""
    print()
    print(f"  ╔═ OpenSeer is asking ═════════════════════════")
    for line in question.splitlines() or [question]:
        print(f"  ║ {line}")
    if kind == "confirm":
        print("  ║ (y/n)")
    elif kind == "choose" and options:
        for i, opt in enumerate(options, 1):
            print(f"  ║  {i}) {opt}")
    print("  ╚══════════════════════════════════════════════")
    try:
        raw = input("  reply > ").strip()
    except EOFError:
        return None
    if not raw:
        return None
    if kind == "confirm":
        return ("Yes" if raw.lower() in ("y", "yes", "是", "ok")
                else "No")
    if kind == "choose" and options:
        if raw.isdigit() and 1 <= int(raw) <= len(options):
            return options[int(raw) - 1]
        # Allow free-text match too.
        for opt in options:
            if opt.lower() == raw.lower():
                return opt
    return raw


def load_rendezvous() -> dict[str, Any]:
    """Read ~/.openseer/agentd.json. Raises FileNotFoundError if
    the daemon hasn't written one (i.e., isn't running)."""
    if not _RENDEZVOUS_PATH.exists():
        raise FileNotFoundError(str(_RENDEZVOUS_PATH))
    return json.loads(_RENDEZVOUS_PATH.read_text())


def try_open(timeout: float = 2.0) -> Optional[AgentdClient]:
    """Best-effort: connect + auth an AgentdClient. Returns a ready-
    to-use client on success, None on any failure (no rendezvous,
    daemon dead, auth fail). Caller MUST call `close()` when done —
    don't `with` it, the connection is already established here.
    Falls back to direct agent.run() if None."""
    try:
        cfg = load_rendezvous()
        client = AgentdClient(cfg["host"], cfg["port"], cfg["token"])
        client.connect_and_auth()
        return client
    except (FileNotFoundError, KeyError, OSError, TimeoutError,
            ConnectionError, ConnectionRefusedError,
            websockets.exceptions.WebSocketException,
            json.JSONDecodeError):
        return None


# ─── rendering helpers (used by the CLI) ────────────────────────────


def render_event_to_stdout(ev: dict[str, Any]) -> None:
    """Translate an event into a terse human-readable line. Matches
    the spirit of agent.run()'s `say()` calls — not byte-identical,
    but enough that running `openseer task` against a live daemon
    feels familiar to someone used to the direct path."""
    et = ev.get("type", "?")
    data = ev.get("data") or {}
    step = ev.get("step")
    if et == "task_started":
        task = data.get("task", "")
        trace = data.get("trace_id", "")
        dry = " (dry_run)" if data.get("dry_run") else ""
        print(f"[agent] task: {task}{dry}")
        print(f"[agent] trace: {trace}")
    elif et == "step_started":
        print(f"\n──── step {step} ────")
    elif et == "model_started":
        pass  # noisy; skip
    elif et == "model_finished":
        usage = data.get("usage") or {}
        elapsed = data.get("elapsed_ms") or 0
        if usage:
            it = usage.get("input_tokens", 0)
            ot = usage.get("output_tokens", 0)
            print(f"  model: {it}+{ot} tokens · {elapsed}ms")
    elif et == "action_started":
        name = data.get("action") or "?"
        print(f"  → {name}")
    elif et == "action_finished":
        result = (data.get("result") or "")[:120]
        if result:
            print(f"    {result}")
    elif et == "step_recorded":
        pass  # already summarised
    elif et == "task_finished":
        status = data.get("status", "done")
        reason = (data.get("reason") or "")[:400]
        print(f"\n[agent] terminated: {status}")
        if reason:
            print(f"        {reason}")
    elif et == "task_failed":
        err = (data.get("error") or "")[:400]
        print(f"\n[agent] crashed: {err}")
    elif et == "agent_held":
        print(f"\n[agent] paused — touch ~/.openseer/runs/<trace>/HOLD to remove and resume")
    elif et == "agent_resumed":
        print(f"[agent] resumed")
